pkg/nsx/services/firewall.go

```diff
@@ -201,45 +218,77 @@ func (service *SecurityPolicyService) buildBasicTags(obj *v1alpha1.SecurityPolic
 	return tags
 }
 
-func (service *SecurityPolicyService) updateTargetExpressions(obj *v1alpha1.SecurityPolicy, target *v1alpha1.SecurityPolicyTarget, group *model.Group, idx int) {
+func (service *SecurityPolicyService) updateTargetExpressions(obj *v1alpha1.SecurityPolicy, target *v1alpha1.SecurityPolicyTarget, group *model.Group, idx int) int {
 	service.appendOperatorIfNeeded(&group.Expression, "OR")
-	expressions := data.NewListValue()
-	expressionFrame := data.NewStructValue(
-		"",
-		map[string]data.DataValue{
-			"expressions":   expressions,
-			"resource_type": data.NewStringValue("NestedExpression"),
-		},
-	)
-	group.Expression = append(group.Expression, expressionFrame)
+	expressions := service.buildGroupExpression(&group.Expression)
 
 	clusterExpression := service.buildExpression(
 		"Condition", "SegmentPort",
 		fmt.Sprintf("%s|%s", util.TagScopeNCPCluster, service.getCluster()),
-		"Tag", "EQUALS",
+		"Tag", "EQUALS", "EQUALS",
 	)
 	expressions.Add(clusterExpression)
+	var err error = nil
+	var podSelectorCount, vMSelectorCount = 0, 0
+
 	if target.PodSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		// TODO: consider to use project_uid instead of project
 		nsExpression := service.buildExpression(
 			"Condition", "SegmentPort",
 			fmt.Sprintf("%s|%s", util.TagScopeNCPProject, obj.ObjectMeta.Namespace),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(nsExpression)
-		service.updatePortExpressions(target.PodSelector.MatchLabels, expressions)
+		service.updateExpressionsMatchLables(target.PodSelector.MatchLabels, "SegmentPort", expressions)
+		if len(target.PodSelector.MatchLabels) > 0 {
+			podSelectorCount = 1
+		}
+
+		if target.PodSelector.MatchExpressions != nil {
+			mergedMatchExpressions := service.mergeSelectorMatchExpression(target.PodSelector.MatchExpressions)
+			podSelectorCount, err = service.validateSelectorExpressions(target.PodSelector.MatchLabels, *mergedMatchExpressions, "SegmentPort")
+			if err == nil {
+				service.updateExpressionsMatchExpression(*mergedMatchExpressions, target.PodSelector.MatchLabels,
+					&group.Expression, clusterExpression, nsExpression, "SegmentPort", expressions)
+			}
+		}
 	}
 	if target.VMSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		nsExpression := service.buildExpression(
 			"Condition", "SegmentPort",
 			fmt.Sprintf("%s|%s", util.TagScopeNCPVIFProject, obj.ObjectMeta.Namespace),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(nsExpression)
-		service.updatePortExpressions(target.VMSelector.MatchLabels, expressions)
+		service.updateExpressionsMatchLables(target.VMSelector.MatchLabels, "SegmentPort", expressions)
+		if len(target.VMSelector.MatchLabels) > 0 {
+			vMSelectorCount = 1
```

Comment:
Same here

Reply:
No, if only vmSelector used, all labelSelectors for vmSelector will produce one group. So VMSelectorCount should be one. for example, 
- vmSelector:
     matchLabels:
          user: internal
          op: role
          user: internal

the above vmSelector will only produce one group, like

<img width="910" alt="image" src="https://user-images.githubusercontent.com/78414502/150449453-d6945525-8b23-4be5-9023-784accb996a8.png">